### ISSUE TITLE
[ranking_factorization_recommender] - Convergence issue when scaling

### ISSUE THREAD

**Issue by benjaminbraunberkeley** [apple/turicreate#1718]
Hi,

I was contacting you because of an issue related to scaling a ranking_factorization_recommender model.

I trained a first version of the model, with 10,000 items around 3 million users. This version was working well, and I could see the error decreasing at every iteration during training.

However, when using the same model on the same type of data but with many more data points (200,000 items),  I see that the model has “difficulties” training; The error rate decreases at some steps, but then the error increases again, the model diverges and a lower learning step is set. 

When creating the training set, all I provided was the (user, item) pairs and some side features of the items. I used a ranking_factorization_recommender, with num_negative_samples = 1. I had a couple of questions related to this problem:

I am not sure of how the negative samples are generated by the package. For instance, is it done by leaving all the columns unchanged and shuffling the user_id column? 

During training, we see the column 'Approx. Training Predictive Error'. How is this error calculated? In my case, I only provided positive examples. So, is this calculated only on the positives, or does it create negative examples and also calculate the error rate on them? 

Also, is it possible to have the training stop not after the maximum number of iterations but when the error is below a certain threshold?

Finally, I wanted to ask for advice on some potential strategies to tackle this scale problem.. Is there a parameter of the 'create' function that you would advise to change? Or do you think it could be a better idea to manually create some negative examples, with negative items chosen in a ’smart’ way (for instance, having some similarity with the positive items chosen)? 

Thanks very much in advance for your answer!

**Comment by hoytak** [comment]
Hey @benjaminbraunberkeley, 

The models with side features have one problem which impacts training.  The reason for this is often due to thread contention on the side features -- multiple sgd updates may clobber each other with the side features, which sometimes causes these models to diverge.  Numeric side features are the worst offenders.  When we designed this algorithm, we considered putting locks around these updates, but in many cases that would have significantly slowed down the training so we did not do that.   

Also, running in single threaded mode would provide a test for this -- this can be done by starting your code with setting the environment variable OMP_NUM_THREADS="1". 

The best way we've found to avoid this scenario is to instead use a large number of categorical features instead of numerical ones.  For example, binning a single numeric feature into the 100 quantiles would help this.  Also, not having side factorization terms but lots of categories may give a good result as well. 

Negative examples are sampled from all the user / item pairs that showed no interaction.  Any user/item side features are attached to those, and the optimization is performed on those.  The sampling strategy, as well as the loss associate with them, follows this paper: https://dl.acm.org/citation.cfm?id=2484126.   Essentially, with a positive user/item data point, it samples a number of items not rated by that user, chooses the item that has the largest value, then does an SGD step on the pair attempting to increase the predicted value of the positive example and decrease the value of the negative example.  Generally we found that more samples give a more stable and better result. 

The Approx. Training Predictive Error is calculated based on the accumulated predictive error of each sample immediately before the SGD descent is provided.  This tracks the true loss quite well. 

There is no way currently to stop after an absolute loss value -- that's a good feature request, though.  You might be able to get by with adjusting the threshold of subsequent loss decrease defined for convergence -- when the loss stops decreasing above a certain rate, the model is assumed to have converged; this is adjustable.

Hope that helps!
-- Hoyt
